src/lib/optimizer/strategy/join_to_predicate_rewrite_rule.cpp

```diff
@@ -0,0 +1,200 @@
+  }
+
+  std::shared_ptr<BinaryPredicateExpression> join_predicate = nullptr;
+  const auto& join_predicates = join_node->join_predicates();
+  for (auto& predicate : join_predicates) {
+    join_predicate = std::dynamic_pointer_cast<BinaryPredicateExpression>(predicate);
+    if (join_predicate) {
+      break;
+    }
+  }
+
+  DebugAssert(join_predicate, "A Join must have at least one BinaryPredicateExpression.");
+
+  std::shared_ptr<AbstractExpression> exchangable_column_expr = nullptr;
+
+  if (expression_evaluable_on_lqp(join_predicate->left_operand(), *removable_subtree)) {
+    exchangable_column_expr = join_predicate->left_operand();
+  } else if (expression_evaluable_on_lqp(join_predicate->right_operand(), *removable_subtree)) {
+    exchangable_column_expr = join_predicate->right_operand();
+  }
+
+  DebugAssert(exchangable_column_expr,
+              "Neither column of the join predicate could be evaluated on the removable input.");
+
+  // Check for uniqueness
+  auto testable_expressions = ExpressionUnorderedSet{};
+  testable_expressions.insert(exchangable_column_expr);
+
+  if (!removable_subtree->has_matching_unique_constraint(testable_expressions)) {
+    return false;
+  }
+
+  // Now, we look for a predicate that can be used inside the substituting table scan node.
+  visit_lqp(removable_subtree, [&removable_subtree, &valid_predicate](auto& current_node) {
+    if (current_node->type == LQPNodeType::Union) return LQPVisitation::DoNotVisitInputs;
+    else if (current_node->type != LQPNodeType::Predicate) return LQPVisitation::VisitInputs;
+
+    const auto candidate = std::static_pointer_cast<PredicateNode>(current_node);
+    const auto candidate_exp = std::dynamic_pointer_cast<BinaryPredicateExpression>(candidate->predicate());
+
+    if (!candidate_exp) {
+      return LQPVisitation::VisitInputs;
+    }
+
+    DebugAssert(candidate_exp, "Need to get a predicate with a BinaryPredicateExpression.");
+
+    // Only predicates that filter using equals cond. on a column with a constant value are of use to our optimization
+    // these conditions have the potential (given filtered column is a UCC) to single out a maximum of one result tuple
+    if (candidate_exp->predicate_condition != PredicateCondition::Equals)
+      return LQPVisitation::VisitInputs;
+
+    auto candidate_column_expr = std::dynamic_pointer_cast<LQPColumnExpression>(candidate_exp->left_operand());
+    std::shared_ptr<ValueExpression> candidate_value_expr = nullptr;
+    if (candidate_column_expr) {
+      candidate_value_expr = std::dynamic_pointer_cast<ValueExpression>(candidate_exp->right_operand());
+    } else {
+      candidate_column_expr = std::dynamic_pointer_cast<LQPColumnExpression>(candidate_exp->right_operand());
+      candidate_value_expr = std::dynamic_pointer_cast<ValueExpression>(candidate_exp->left_operand());
+    }
+
+    // should not have a case where we run into no column, but may be a case where we have no value but compare columns
```

Comment:
overall feedback for your comments: the hyrise code tries to use grammatically correct sentences that start with an uppercase letter and end with a full stop wherever possible.